http-verbs-common/src/main/scala/uk/gov/hmrc/http/client2/HttpClient2.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.http.client2
+
+import play.api.libs.ws.{BodyWritable, WSRequest}
+import uk.gov.hmrc.http.{HeaderCarrier, HttpReads}
+
+import java.net.URL
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.runtime.universe.TypeTag
+
+/** This client centralises the execution of the request to ensure that the common concerns (e.g. auditing, logging,
+  * retries) occur, but makes building the request more flexible (by exposing play-ws).
+  * It also supports streaming.
+  */
+trait HttpClient2 {
+  protected def mkRequestBuilder(url: URL, method: String)(implicit hc: HeaderCarrier): RequestBuilder
+
+  def get(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "GET")
+
+  def post(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "POST")
+
+  def put(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "PUT")
+
+  def delete(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "DELETE")
+
+  def patch(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "PATCH")
+
+  def head(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "HEAD")
+
+  def options(url: URL)(implicit hc: HeaderCarrier): RequestBuilder =
+    mkRequestBuilder(url, "OPTIONS")
+}
+
+trait RequestBuilder {
+  def transform(transform: WSRequest => WSRequest): RequestBuilder
```

Comment:
Have you considered not exposing `transform` on the basis that it might be too powerful (e.g. the `transform(_.withBody)` issue)? Or are there too many use-cases for us to anticipate without providing it?

Reply:
This is essentially the change in approach from `HttpClient` which tried to encapsulate `play-ws`, and caused problems whenever a client needed to make a change not exposed (e.g. change the request timeout). Rather than overload for every feature supported by `play-ws`, it seems more scalable to just expose the `WSRequest`.

Reply:
Makes sense, thanks for the context